ocaml/gencert/validation.ml

```diff
@@ -51,58 +51,197 @@ let validate_private_key pkcs8_private_key =
-      | Ok (cert :: _) ->
-          Ok cert
-      | Ok [] ->
-          D.info "Rejected certificate chain because it's empty." ;
-          Error (`Msg (server_certificate_chain_invalid, []))
-      | Error (`Msg err_msg) ->
-          D.info {|Failed to validate certificate chain because "%s"|} err_msg ;
-          Error (`Msg (server_certificate_chain_invalid, []))
+  ensure_validity ~time server >>= ensure_sha256_signature_algorithm
+
+let validate_server_certificate cert time private_key =
+  let ensure_keys_match private_key certificate =
+    let public_key = X509.Certificate.public_key certificate in
+    match (public_key, private_key) with
+    | `RSA pub, `RSA priv when pub = Mirage_crypto_pk.Rsa.pub_of_priv priv ->
+        Ok ()
+    | _ ->
+        Error (`Msg (server_certificate_key_mismatch, []))
+  in
+  ensure_keys_match private_key cert >>= fun () ->
+  validate_certificate cert time server_certificate_not_valid_yet
+    server_certificate_expired
+
+let validate_intermediate_certificate cert time =
+  validate_certificate cert time server_certificate_intermediate_not_valid_yet
+    server_certificate_intermediate_expired
+
+let issuer_matches_subject parent cert =
+  X509.(
+    Distinguished_name.equal
+      (Certificate.subject parent)
+      (Certificate.issuer cert))
+
+let ext_authority_matches_subject trusted cert =
+  match
+    X509.(
+      Extension.
+        ( find Authority_key_id (Certificate.extensions cert)
+        , find Subject_key_id (Certificate.extensions trusted) ))
+  with
+  | _, None | None, _ ->
+      (* not mandatory *)
+      true
+  | Some (_, (Some auth, _, _)), Some (_, au) ->
+      Cstruct.equal auth au
+  (* TODO: check exact rules in RFC5280 *)
+  | Some (_, (None, _, _)), _ ->
+      (* not mandatory *)
+      true
+
+let signs pathlen trusted cert =
+  match
+    ( issuer_matches_subject trusted cert
+    , ext_authority_matches_subject trusted cert
+    , X509.Validation.validate_signature supported_cert_hashes trusted cert
+    , X509.Validation.validate_path_len pathlen trusted )
+  with
+  | true, true, Ok (), true ->
+      Ok ()
+  | _ ->
+      Error (`Msg (server_certificate_chain_invalid, []))
+
+let rec validate_anchors pathlen cert = function
+  | [] ->
+      Error (`Msg (server_certificate_chain_invalid, []))
+  | x :: xs -> (
+    match signs pathlen x cert with
+    | Ok _ ->
+        Ok x
+    | Error _ ->
+        validate_anchors pathlen cert xs
+  )
+
+let issuers_of trusted cert =
+  List.filter (fun p -> issuer_matches_subject p cert) trusted
+
+let verify_single_chain time anchors chain =
+  let open X509.Validation in
+  let rec climb pathlen = function
+    | cert :: issuer :: certs ->
+        validate_intermediate_certificate issuer time >>= fun _ ->
+        signs pathlen issuer cert >>= fun () ->
+        climb (succ pathlen) (issuer :: certs)
+    | [c] ->
+        let anchors = issuers_of anchors c in
```

Comment:
It's not obvious whether these anchors have are 'valid'. I believe they are because of `List.filter (has_valid_time time) anchors` below, but it would be simpler to do this check / filtering inside `load_ca_cert_bundle`